packages/vm/lib/runTx.ts

```diff
@@ -243,6 +243,17 @@ async function _runTx(this: VM, opts: RunTxOpts): Promise<RunTxResult> {
         `sender doesn't have enough funds to send tx. The upfront cost is: ${cost} and the sender's account only has: ${balance}`
       )
     }
+    if (tx instanceof FeeMarketEIP1559Transaction) {
+      // EIP-1559 spec:
+      // The signer must be able to afford the transaction
+      // `assert balance >= gas_limit * max_fee_per_gas`
+      const cost = tx.gasLimit.mul(tx.maxFeePerGas)
+      if (balance.lt(cost)) {
+        throw new Error(
+          `sender doesn't have enough funds to send tx. The max cost is: ${cost} and the sender's account only has: ${balance}`
+        )
+      }
+    }
```

Comment:
@ryanio Hmm, when having a second look here, I guess we should rather adopt the 1559 tx [getUpfrontCost()](https://github.com/ethereumjs/ethereumjs-monorepo/blob/master/packages/tx/src/eip1559Transaction.ts#L239) implementation instead of doing this just-one-tx-type check here in *addition* to the `getUpfrontCost()` check (above in the code). 🤔

This is what the `getUpfrontCost()` docs say it should do: "The up front amount that an account must have for this transaction to be valid"

Reply:
(or are these two different things?)

Reply:
@holgerd77 under its current implementation it is a different check, since `getUpfrontCost` uses the `inclusionFeePerGas ` and also adds the tx's `value`:

```typescript
getUpfrontCost(baseFee: BN = new BN(0)): BN {
    const inclusionFeePerGas = BN.min(this.maxPriorityFeePerGas, this.maxFeePerGas.sub(baseFee))
    const gasPrice = inclusionFeePerGas.add(baseFee)
    return this.gasLimit.mul(gasPrice).add(this.value)
  }
```

compared to the 1559 validation check which simply does:

```typescript
const cost = tx.gasLimit.mul(tx.maxFeePerGas)
```

Example tx with:
  `gasLimit=21000, maxFeePerGas=100, maxPriorityFeePerGas=10, value=100, block.baseFeePerGas=7`

getUpfrontCost: `21000*(10+7)+100=357100`
1559 validation max cost check: `21000*100=2100000`

that said, we could change `getUpfrontCost` to the new validation logic, but it seems like it could be nice to have both, @jochem-brouwer wdyt?

Reply:
Yeah, might be good to just keep both, thanks for these concrete examples.